### DISCUSSION TITLE
3.7.4 - Var's parent node isn't the StatementPattern in which is appears (after parsing query with property paths)?

### DISCUSSION THREAD

**Discussion by tayloj** [eclipse-rdf4j/rdf4j#4645] [General]
This query parses as follows (and I've labeled a few nodes to facilitate discussion):

```sparql
prefix : <http://example.org/>
select * { ?s :f/:g ?o }
```

```
Projection
   ProjectionElemList
      ProjectionElem "s"
      ProjectionElem "o"
   Join
      StatementPattern  # lhs
         Var (name=s)
         Var (name=_const_c0e515a5_uri, value=http://example.org/f, anonymous)
         Var (name=_anon_6b8ae105_4707_48d9_957f_6b2c7c760218, anonymous)  # lhsX
      StatementPattern  # rhs
         Var (name=_anon_6b8ae105_4707_48d9_957f_6b2c7c760218, anonymous) # rhsX
         Var (name=_const_c0e515a6_uri, value=http://example.org/g, anonymous)
         Var (name=o)
```

The parent of rhsX is rhs, which is what I'd expect, but the parent of lhsX is not lhs (and also is not rhs), but is a statement pattern that doesn't appear in the over parsed query.

This means that from an arbitrary node in the query, it's not possible to use `getParentNode()` to walk back up to the root of the parsed query.  Any ideas what's going on here?  I'm using RDF4j 3.7.4 (dictated by external factors, but if the behavior is different in later versions, that's still valuable information).

Here's a JUnit that demonstrates the behavior:

```java
   @Test
   public void statementsWithoutParents() {
      final ParsedQuery q = new SPARQLParser().parseQuery("prefix : <http://example.org/> select * { ?s :f/:g ?o }", null);
      final Projection projection = (Projection) q.getTupleExpr();
      final Join join = (Join) projection.getArg();
      final StatementPattern left = (StatementPattern) join.getLeftArg();
      final StatementPattern right = (StatementPattern) join.getRightArg();
      final Var leftX = left.getObjectVar();
      final Var rightX = right.getSubjectVar();
      assertThat(rightX.getParentNode(), is(right)); // succeeds
      assertThat(leftX.getParentNode(), is(left)); // fails
      //
      //      java.lang.AssertionError:
      //         Expected: is <StatementPattern
      //            Var (name=s)
      //            Var (name=_const_c0e515a5_uri, value=http://example.org/f, anonymous)
      //            Var (name=_anon_2134e1c5_851c_476a_bbc1_33102b7e701d, anonymous)
      //         >
      //              but: was <StatementPattern
      //            Var (name=_anon_2134e1c5_851c_476a_bbc1_33102b7e701d, anonymous)
      //            Var (name=_const_c0e515a6_uri, value=http://example.org/g, anonymous)
      //            Var (name=_anon_43b2b107_38d3_41d4_b72c_41ffe5ba0415, anonymous)
      //         >
   }
```

As a workaround, since I'm actual using the visitor in real code (not manually extracting the bits as in the JUnit), I've just set the parent node on the children of statement patterns before calling `super.meet(node)`,  and that seems to patch things up for me.  Is there any danger in this approach?

```java
         @Override
         public void meet(final StatementPattern node) throws RuntimeException {
            node.getSubjectVar().setParentNode(node);
            node.getPredicateVar().setParentNode(node);
            node.getObjectVar().setParentNode(node);
            super.meet(node);
// ...
```

**Comment by hmottestad** [comment]
We fixed this at some point, not sure exactly when but it definitely shouldn't be a problem on the latest version. We also implemented a global test for this in https://github.com/eclipse/rdf4j/blob/main/core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/optimizer/ParentReferenceChecker.java .

**Reply by tayloj** [reply]
@hmottestad Thanks, it's good to hear that's been fixed in a later version! I'm kind of pinned to 3.7.4 for the moment (existing application with dependencies I can't change, etc.); does the workaround with setting the parent of a StatementPattern's children seem appropriate, or is there any danger that I might not have anticipated?

**Reply by hmottestad** [reply]
One issue I found was that the Var object was being shared between multiple StatementPattern objects. This cropped up 2-3 times in the code. Might not be the case with your query, but I wouldn't trust the parent ref if I were you.

If you need to keep track of the parent you can take a look at the ParentCheckingVisitor n the link I provided. It uses a deque to keep track of the parent object.
